Make `Expr` untyped

joneshf-cn opened the pull request:
##### What it does

Tries an approach to encoding expressions as untyped. The motivation behind this change is that the well-typed expressions were giving us issues when attempting to evaluate them. If we have no real types, we can always attempt to evaluate something. It ends up that we have to deal with failure in the cases where we've constructed an ill-typed expression, though.

##### Review asks

Is this approach something we want to continue pursuing? I'm more than happy to close this PR in exchange for something else, but I'm not sure what that something else might be.

There are a few open questions that I couldn't figure out on my own. I left them as review notes in this PR.

joneshf-cn committed:
Try a more dynamically typed approch

The well typed approach seems to be giving us some problems.
When we want to add lookup, we start running into typed issues.
Surely there's a way around it.

In the meantime, we explore an approach with more relaxed types.
We don't statically verify any of the expressions are well formed.
We dynamically check whether they can work.

This isn't the most ideal approach,
as we now have to deal with possible failures in multiple places.
However, it's forward progress.
We can also move to a more well-typed approach incrementally.

There are still ways we can make this better.
We might be able to type check at the edges,
and construct a better typed expression.
We might transform the failures into usable output.
We might do something else.

In any case, we now have some dynamic behavior in these forms!

joneshf-cn committed:
Get tests passing

We killed the decoders/encoders while experimenting with this approach.
Now that we've seen it has legs, we put them back so we can parse JSON.

joneshf-cn committed:
Re-add printing

This was removed as part of the experiment.
But, we can add it back now that we know this idea can work.

joneshf-cn commented on src/Data/Expr.purs:
```diff
@@ -4,259 +4,157 @@ import Prelude
+  If x y z -> IfA (toExprA x) (toExprA y) (toExprA z) (reflectIn x) (reflectOut y)
+  Equal x y -> EqualA (toExprA x) (toExprA y) (reflectIn x) (reflectOut x)
+  Print x -> PrintA (toExprA x) (reflectIn x) "String"
+  Lookup x -> LookupA x "???"
```
❓ What do we do here? Does `LookupA` have an output type? Is it `Unit`? Is it something else?

davezuch replied:
I think this should be the same type of whatever value you're looking up. Which should be known at encode time, since each input type has only one valid type. Also, there are type constraints on where in an expression it's used. 

Some clarifying examples:
```purescript
{ name: string_ "Name"
, visibility: lookup_ "active"
, description: if_ (lookup_ "active") (string_ "you need a name") (lookup_ "foo")
}
```
So above,  `visibility` needs to be a Boolean, so only boolean lookups are valid here, so this could only have been encoded if `active` is a Boolean input. Similarly, in `description`, the `active` lookup is valid because an `if_` condition has to be a Boolean expression.

joneshf-cn replied:
Hmm, it seems like we'd have to have the `Page Expr` then to do that, right? Or at least some `Map Key String`/`Key -> String` or whatever. The `Key`s only exist in a `Page Expr`, so maybe we can't have an instance for `EncodeJson Expr`. It would have to be `EncodeJson (Page Expr)` and friends, so they could traverse with the type context.

That doesn't seem too bad, since `ExprA`s reason for existence seems to be to facilitate encoding.

davezuch replied:
I see what you mean, with `ExprF i o` we never lose our output type but if we drop our types then we have no way of recovering that information without additional context.

One trick we could try would be to have a failure case for our lookup, something like
```purescript
  | Lookup Key Expr
```
That way we can kind of cheat and look at the output type of the failure case Expr.

joneshf-cn replied:
That seems like it'll also remove one more error case to deal with. Lemme give it a shot.

joneshf-cn replied:
That seemed to work out! Great suggestion!

joneshf-cn commented on src/Data/Expr.purs:
```diff
@@ -4,259 +4,157 @@ import Prelude
-  toExprType = Boolean
-  reflectProxy _ = "Boolean"
-  print = show
+print :: ExprType -> ExprType
```
I think this is typed not right, it was thrown together when adding `Print` back in the mix, but it looks like we probably want the type to be `ExprType -> String` and we can do the wrapping where it's used.

joneshf-cn commented on src/Data/Expr.purs:
```diff
@@ -4,259 +4,157 @@ import Prelude
+evalExpr :: (Key -> Maybe ExprType) -> Expr -> Either EvalError ExprType
+evalExpr get = case _ of
+  Val x -> pure x
+  If x' y z -> case evalExpr get x' of
```
I think the explicit casing on the `Either _ _` can be replaced with binds. Maybe something like:

```PureScript
If x' y z -> do
  x <- evalExpr get x'
  case x of
    Boolean false -> evalExpr get z
    Boolean true -> evalExpr get y
    _ -> Left (IfCondition x)
```
or
```PureScript
If x' y z -> evalExpr get x' >>= case _ of
  Boolean false -> evalExpr get z
  Boolean true -> evalExpr get y
  x -> Left (IfCondition x)
```

Not a huge deal on this one, but it might clean up the `Equal` case a bit.

joneshf-cn commented on src/Data/Form.purs:
```diff
@@ -138,23 +58,18 @@ type InputState t r =
+  | Toggle (Record (SharedRows f ()))
 
-derive instance genericInput :: Generic (Input f) _
+derive instance genericInput :: Generic (Input Expr) _
```
I changed this on accident. Maybe we can remove the change so the diff is a little cleaner?

```suggestion
derive instance genericInput :: Generic (Input f) _
```

joneshf-cn commented on src/Data/Form.purs:
```diff
@@ -138,23 +58,18 @@ type InputState t r =
     Toggle r -> "type" := "Toggle" ~> encodeJson r
 
-instance decodeInput :: DecodeJson (Input Expr) where
+instance decodeJsonInput :: DecodeJson (Input Expr) where
```
Similar accidental change.

```suggestion
instance decodeInput :: DecodeJson (Input Expr) where
```

joneshf-cn commented on src/Data/Form.purs:
```diff
@@ -243,15 +158,19 @@ lastName =
     }
   }
+  where
+  description = if_ (equal_ (boolean_ true) (lookup_ "active"))
```
I wonder if we can drop the `equal_` stuff? It should be semantically the same as `if_ (lookup_ "active")`, right?
```suggestion
  description = if_ (lookup_ "active")
```

joneshf-cn commented on src/Page/Form.purs:
```diff
@@ -19,12 +24,14 @@ import Ocelot.Block.Toggle (toggle) as Toggle
 
 type State =
-  { form :: Either String (Page Expr)
+  { form :: Either String { expr :: Page Expr, evaled :: Page ExprType }
```
I'm not sold on the naming here. Happy to change it to whatever if we've got better names. Also fine to keep it as is, if we don't have anything better.

davezuch replied:
This is better than anything I came up with when I was trying lol

joneshf-cn commented on src/Page/Form.purs:
```diff
@@ -47,18 +54,105 @@ component =
+        values = keysPage expr
+    H.modify_ _ { values = values }
+    case evaled' of
+      Left _ -> pure unit
```
❓ This ignores any errors that we had when we're evaluating the form. What should we do in this case? Do we want to throw the message up in a toast or something? Do we log errors somewhere we can later take a look at?

davezuch replied:
At this point an error is fatal, so I think we could replace the whole `case` with
```purescript
H.modify_ _ { form = { expr, evaled: _ } <$> evaled }
```
And update the type of our state so the `Left` is an actual error type

joneshf-cn commented on src/Page/Form.purs:
```diff
@@ -47,18 +54,105 @@ component =
+  keysField :: Field Expr -> Map Key ExprType
+  keysField field = case field.input of
+    Text { value: { value: Just expr } }
+      | Right value <- evalExpr (const Nothing) expr ->
```
❓ These `const Nothing`s are actually limiting the amount of depth we're able to have for lookups. If the expression we're looking up is a lookup itself, it won't find the key, even if it exists. Maybe there's a different way to do this? Maybe we should ditch the idea of a map in favor of the pub/sub stuff?

davezuch replied:
Hmm, I'm not sure our lookups need to be recursive. I mean, that would be pretty fancy, but the results of our lookups will be the value of a form field, and I'm not really imagining allowing users to build up expressions in the generated form. Although...that would be pretty mind blowing 🤔

joneshf-cn replied:
Oh right, that makes sense! The thing we're looking up is an `ExprType` anyway, not an `Expr`. Sorry about that.

joneshf-cn reviewed (commented):
Giving myself a once-over for to-dos so I don't forget them.

I'll address what I can and leave the unaddressed stuff for review.

joneshf-cn committed:
Apply suggestions from code review

Some of this is to make the overall diff cleaner.

The changes to `description` are to simplify it a bit.

joneshf-cn committed:
Turn warnings into errors

As with a previous commit,
if we want to get lots of signal from the compiler,
we should mitigate the amount of noise we let through.

If we turn the warnings into errors, CI will catch these and fail on us.
We can address them quicker with a short feedback loop like this
and keep the compiler telling us useful things.

joneshf-cn committed:
Clean up some code

This is should be all refactorings, no behavioral changes.

joneshf-cn committed:
Fix npm scripts

We cannot say `pulp build --to` and also pass options to `psa`.
So, we have to take it in two steps.

Maybe there's some incantation that actually supports one line.
Until then, this works.

davezuch commented on src/Page/Form.purs:
```diff
@@ -47,18 +54,105 @@ component =
+      Right evaled -> H.modify_ _ { form = pure { expr, evaled } }
+
+  eval (UpdateKey key val a) = do
+    form <- H.gets _.form
```
This is a tiny, mostly cosmetic thing, but I like to write these as
```purescript
{ form } <- H.get
```
I think it's a little cleaner, and allows you to match multiple values more efficiently if you needed to do something like
```purescript
{ form, values } <- H.get
```

joneshf-cn committed:
Use the error value as an actual error

If we've evaluated something and it ends up being an error,
we've broken some important invariant.
Which is to say, this error isn't recoverable.
We throw our hands up, display the error, and let someone else handle it.

joneshf-cn committed:
Require a default value for lookup

Since we've thrown away our types,
we don't have the information about what the lookup was going to be.
We take along with us a proxy for what the value should be.

We can use this proxy as a default value for the lookup.
This has two benefits:
1. We can now infer what the type of the lookup should be.
2. We can always have _some_ value when looking up.
    If lookup fails, we can return the default value.
    This means we can remove one of our error cases!

joneshf-cn committed:
Use an `Expr` as the default instead of `ExprType`

There's no reason we have to restrict the default value to `ExprType`.
If someone wants to use an `Expr` as a default value, they can.

There was a concern that evaluation wouldn't terminate,
but `Expr` is inductive data.
Evaluation will terminate at some point.

This allows for a bit of choice when constructing lookups.
For instance:
```PureScript
lookup_ "foo" (lookup_ "bar" (lookup_ "baz" (boolean_ true)))
```

We check for three different keys in order,
and finally default to `true`.
Everything is still total, and all works well.

joneshf-cn commented:
I think I addressed all the comments. Lemme know if there's more we should do here. If not, I'll merge this after CI passes.

davezuch reviewed (approved):
I'm happy with this. Honestly this change to dynamic typing seems quite promising, at the least it's easier to work with.
